snapcraft/internal/steps.py

```diff
@@ -0,0 +1,97 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2018 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from snapcraft.internal import errors
+
+
+class Step:
+    def __init__(self, name: str) -> None:
```

Comment:
I am just going to throw this out, but what if instead of keeping a tight coupling with `STEPS` and those `index` calls to know where it lives we create the Step instance with knowledge of the next step and the previous step, kind of like a linked list and create a map of step names to the instances?

Reply:
Yeah I thought about that, but ended up having two issues:

1. Order needs to be provided somehow, whether that's by providing a list giving the order or by hard-coding it into each `Step` instance. The former seemed simpler to understand and modify than the latter.
2. We need a way to know which step is first. The `STEPS` variable provides that knowledge (STEPS[0]).

It's an interesting idea, though. Let me try it real quick, I think it's a small change. Then we can see what we think.

Reply:
Alright, just pushed it up. Let me know what you think.

Reply:
We discussed offline, and I've moved it back to a list, but kept the same methods.